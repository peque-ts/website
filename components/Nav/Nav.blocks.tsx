import clsx from 'clsx';
import Link from 'next/link';
import React from 'react';

import { NAV_ITEMS, useNavMobileButton } from './Nav.helpers';

interface NavMobileProps {
  bgClassName?: string;
}

const NavMobile: React.VFC<NavMobileProps> = ({ bgClassName }) => {
  const { open, renderButton } = useNavMobileButton();

  return (
    <div className="tablet:hidden block">
      {renderButton()}
      <div
        className={clsx(
          'absolute inset-x-0 -z-10 shadow-lg transition',
          bgClassName,
          open ? 'translate-y-0' : '-translate-y-full',
        )}
      >
        {open && (
          <ul className="flex flex-col space-y-4 items-center pt-4 pb-8">
            {NAV_ITEMS.map(({ name, to }, index) => (
              <li key={index}>
                <Link href={to}>
                  <a className="text-lg">{name}</a>
                </Link>
              </li>
            ))}
          </ul>
        )}
      </div>
    </div>
  );
};

const NavDesktop: React.VFC = () => (
  <ul className="hidden tablet:flex space-x-8">
    {NAV_ITEMS.map(({ name, to }, index) => (
      <li key={index}>
        <Link href={to}>
          <a className="text-lg">{name}</a>
        </Link>
      </li>
    ))}
  </ul>
);

export { NavMobile, NavDesktop };
