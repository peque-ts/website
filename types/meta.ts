export interface ProjectMeta {
  title: string;
  description: string;
}

export interface SectionMeta {
  title: string;
  description: string;
  order: number;
}
