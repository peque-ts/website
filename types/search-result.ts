export interface SearchResult {
  title: string;
  description: string;
  link: string;
}
