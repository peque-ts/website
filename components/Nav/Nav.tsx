import React from 'react';

import { NavDesktop, NavMobile } from './Nav.blocks';

interface Props {
  bgClassName?: string;
}

export const Nav: React.VFC<Props> = ({ bgClassName }) => (
  <nav>
    <NavMobile bgClassName={bgClassName} />
    <NavDesktop />
  </nav>
);
