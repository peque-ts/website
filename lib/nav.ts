import matter from 'gray-matter';
import fs from 'node:fs/promises';
import path from 'node:path';

import { read } from './fs';
import { NavItem, PrevNextNavItems } from './nav.types';

async function buildSideNavItems(project: string, activeSection: string): Promise<NavItem[]> {
  const items: NavItem[] = [];

  const sectionFiles = await fs.readdir(path.join(process.cwd(), 'docs', project));

  for (const sectionFile of sectionFiles) {
    const markdown = await read(`docs/${project}/${sectionFile}`);

    const sectionSlug = sectionFile.replace(/.md$/, '');
    const { title, order } = matter(markdown).data;

    items.push({
      name: title,
      to: `/docs/${project}/${sectionSlug}`,
      order,
      active: sectionSlug === activeSection,
      slug: sectionSlug,
    });
  }

  return items.sort((a, b) => {
    return a.order > b.order ? 1 : a.order < b.order ? -1 : 0;
  });
}

function getPrevNextNavItems(sideNavItems: NavItem[], activeSection: string): PrevNextNavItems {
  const currentItem = sideNavItems.find((item) => item.slug === activeSection);

  if (!currentItem) {
    throw new Error(
      `Cannot find navigation item with slug = [${activeSection}]. This looks like a bug.`,
    );
  }

  return {
    prev: sideNavItems.find((item) => item.order === currentItem.order - 1) ?? null,
    next: sideNavItems.find((item) => item.order === currentItem.order + 1) ?? null,
  };
}

export { buildSideNavItems, getPrevNextNavItems };
