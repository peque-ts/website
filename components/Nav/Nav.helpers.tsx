import clsx from 'clsx';
import React, { useCallback, useState } from 'react';

import { NavItem } from '../../lib/nav.types';

export const NAV_ITEMS: Pick<NavItem, 'name' | 'to'>[] = [
  {
    name: 'Docs',
    to: '/docs',
  },
  {
    name: 'Enterprise',
    to: '/enterprise',
  },
  {
    name: 'Sponsors',
    to: '/sponsors',
  },
];

export const useNavMobileButton = () => {
  const [open, setOpen] = useState(false);

  const renderButton = useCallback(() => {
    const renderSegment = (className: string) => (
      <span
        aria-hidden="true"
        className={clsx(
          'block absolute h-0.5 w-5 bg-white transform transition ease-in-out',
          className,
        )}
      />
    );

    return (
      <button
        type="button"
        className="w-10 h-10 relative focus:outline-none"
        onClick={() => setOpen((value) => !value)}
      >
        <span className="sr-only">Open main menu</span>
        <div className="block w-5 absolute left-1/2 top-1/2 transform -translate-x-1/2 -translate-y-1/2">
          {renderSegment(open ? 'rotate-45' : '-translate-y-1.5')}
          {renderSegment(open ? 'opacity-0' : '')}
          {renderSegment(open ? '-rotate-45' : 'translate-y-1.5')}
        </div>
      </button>
    );
  }, [open]);

  return { renderButton, open };
};
