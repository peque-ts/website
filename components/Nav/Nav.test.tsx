import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { NavItem } from '../../lib/nav.types';
import { Nav } from './Nav';

describe('Nav', () => {
  const items: NavItem[] = [
    {
      active: true,
      name: 'First',
      to: '/docs/first/sub-1',
      slug: 'first/sub-1',
      children: [
        {
          active: true,
          name: 'Sub 1',
          to: '/docs/first/sub-1',
          slug: 'first/sub-1',
          children: [],
        },
        {
          active: false,
          name: 'Sub 2',
          to: '/docs/first/sub-2',
          slug: 'first/sub-2',
          children: [],
        },
      ],
    },
    {
      active: false,
      name: 'Second',
      to: '/docs/2',
      slug: 'second',
      children: [],
    },
    {
      active: false,
      name: 'Third',
      to: '/docs/3',
      slug: 'third',
      children: [],
    },
  ];

  it('should render', () => {
    const { asFragment } = render(<Nav items={items} />);

    expect(screen.getAllByRole('link')).toHaveLength(4);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should toggle mobile nav', () => {
    render(<Nav items={items} />);

    // assert hidden by default
    expect(screen.getAllByRole('list')[0].parentElement).toHaveClass('hidden');

    // expand menu
    userEvent.click(screen.getAllByRole('button')[0]);

    // assert no longer hidden
    expect(screen.getAllByRole('list')[0].parentElement).not.toHaveClass('hidden');

    // collapse menu
    userEvent.click(screen.getAllByRole('button')[0]);

    // assert hidden again
    expect(screen.getAllByRole('list')[0].parentElement).toHaveClass('hidden');
  });

  it('should toggle submenu', () => {
    render(<Nav items={items} />);

    expect(screen.getAllByRole('list')).toHaveLength(2);

    userEvent.click(screen.getByText('First'));

    expect(screen.getAllByRole('list')).toHaveLength(1);

    userEvent.click(screen.getByText('First'));

    expect(screen.getAllByRole('list')).toHaveLength(2);
  });
});
