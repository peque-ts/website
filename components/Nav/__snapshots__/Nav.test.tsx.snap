// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should render 1`] = `
<DocumentFragment>
  <nav>
    <div
      class="flex justify-center laptop:hidden"
    >
      <button
        class="flex items-center space-x-2 px-8"
        type="button"
      >
        <span
          class="text-sm font-semibold"
        >
          Menu
        </span>
        <span
          class="transition ease-in-out -rotate-90"
        >
          <svg
            height="12"
            viewBox="0 0 64 64"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.15.5 4 6.65a.48.48 0 0 0 0 .7l6.15 6.15"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              transform="rotate(.045) scale(4.57143)"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="laptop:block hidden"
    >
      <ul>
        <li>
          <div>
            <button
              class="flex items-center space-x-2 py-2 text-secondary-200 transition hover:text-white text-white font-semibold"
              type="button"
            >
              <span
                class="flex transition ease-in-out -rotate-90"
              >
                <svg
                  height="8"
                  viewBox="0 0 64 64"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10.15.5 4 6.65a.48.48 0 0 0 0 .7l6.15 6.15"
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    transform="rotate(.045) scale(4.57143)"
                  />
                </svg>
              </span>
              <span>
                First
              </span>
            </button>
            <ul
              class="pl-4"
            >
              <li>
                <a
                  class="flex py-2 text-secondary-200 transition hover:text-white font-semibold text-white"
                  href="/docs/first/sub-1"
                >
                  Sub 1
                </a>
              </li>
              <li>
                <a
                  class="flex py-2 text-secondary-200 transition hover:text-white"
                  href="/docs/first/sub-2"
                >
                  Sub 2
                </a>
              </li>
            </ul>
          </div>
        </li>
        <li>
          <a
            class="flex py-2 text-secondary-200 transition hover:text-white"
            href="/docs/2"
          >
            Second
          </a>
        </li>
        <li>
          <a
            class="flex py-2 text-secondary-200 transition hover:text-white"
            href="/docs/3"
          >
            Third
          </a>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;
